### PR TITLE
INN-2310: Implement step.sendEvent

### PR DESCRIPTION
**Pull request by KiKoS0**
Add support for [`step.sendEvent`](https://www.inngest.com/docs/reference/functions/step-send-event) to send events out.

Introduces a small refactor to the `CommHandler` http client to dry out API calls.

### CODE REVIEW

File: inngest-core/src/main/kotlin/com/inngest/Step.kt [hunk 1]
```diff
@@ -77,4 +84,43 @@ class Step(val state: State) {
...
+        id: String,
+        events: Array<InngestEvent>,
+    ) {
+        val hashedId = state.getHashFromId(id)
+
+        try {
+            // If this doesn't throw an error, it's null and that's what is expected
+            val stepState = state.getState<Any?>(hashedId)
+            if (stepState != null) {
+                throw Exception("step state expected sleep, got something else")
```
Comment (darwin67) [inline inngest-core/src/main/kotlin/com/inngest/Step.kt:118]: probably not expecting sleep here.

File: inngest-core/src/main/kotlin/com/inngest/Step.kt [hunk 2]
```diff
@@ -77,4 +84,43 @@ class Step(val state: State) {
...
+     * Sends a single event to Inngest
+     *
+     * @param id Unique step id for memoization.
+     * @param events An array of event payload objects.
+     *
+     */
+    fun sendEvent(
+        id: String,
+        events: Array<InngestEvent>,
+    ) {
```
Comment (darwin67) [inline inngest-core/src/main/kotlin/com/inngest/Step.kt:111]: the return type should be similar to `send` itself.
https://www.inngest.com/docs/reference/functions/step-send-event#step-send-event-id-event-payload-event-payload-promise-ids-string

**Review by darwin67 — APPROVED** [review]
non blocking comments.
if you're not planning to address them in this PR, would be best to add some `TODO` comments so it's clear what needs to be revisit later on.

File: inngest-core/src/main/kotlin/com/inngest/Step.kt (hunk 2 above)
Comment (KiKoS0) [reply inngest-core/src/main/kotlin/com/inngest/Step.kt:111]: gotcha, I'll address them in this PR 👍

File: inngest-core/src/main/kotlin/com/inngest/Function.kt [hunk 3]
```diff
@@ -19,7 +19,8 @@ data class FunctionTrigger(
 enum class OpCode {
     StepRun,
     Sleep,
-    StepStateFailed, // TODO
+    StepStateFailed,
```
Comment (albertchae) [inline inngest-core/src/main/kotlin/com/inngest/Function.kt:22]: still TODO right?

File: inngest-core/src/main/kotlin/com/inngest/Step.kt [hunk 4]
```diff
@@ -77,4 +84,43 @@ class Step(val state: State) {
...
+        try {
+            // If this doesn't throw an error, it's null and that's what is expected
+            val stepState = state.getState<Any?>(hashedId)
+            if (stepState != null) {
+                throw Exception("step state expected sleep, got something else")
+            }
+            return
+        } catch (e: StateNotFound) {
+            val response = CommHandler.sendEvent<SendEventsResponse>(events)
+            throw StepInterruptSendEventException(id, hashedId, response!!.ids)
```
Comment (albertchae) [inline inngest-core/src/main/kotlin/com/inngest/Step.kt:123]: minor, let's see if we can change the return type of `CommHandler.sendEvent` to guarantee non null response

File: inngest-core/src/main/kotlin/com/inngest/Step.kt (hunk 1 above)
Comment (albertchae) [reply inngest-core/src/main/kotlin/com/inngest/Step.kt]: edited to say sendEvent and clarified TODO in State.kt to check for sendEvent as well

File: inngest-core/src/main/kotlin/com/inngest/Step.kt (hunk 2 above)
Comment (albertchae) [reply inngest-core/src/main/kotlin/com/inngest/Step.kt:111]: Turns out this is a bit tricky, we left a TODO to investigate later

File: inngest-core/src/main/kotlin/com/inngest/Step.kt (hunk 4 above)
Comment (albertchae) [reply inngest-core/src/main/kotlin/com/inngest/Step.kt:125]: we'll investigate this later too

**Review by albertchae — APPROVED** [review]

**Pull request merged**